lib/spack/spack/cmd/find.py

```diff
@@ -299,43 +297,66 @@ def root_decorator(spec, string):
         print()
 
 
-def find(parser, args):
-    env = ev.active_environment()
-
-    if not env and args.only_roots:
-        tty.die("-r / --only-roots requires an active environment")
-    if not env and args.show_concretized:
-        tty.die("-c / --show-concretized requires an active environment")
-
+def _find_query(args, env):
+    q_args = query_arguments(args)
+    concretized_but_not_installed = list()
     if env:
+        all_env_specs = env.all_specs()
         if args.constraint:
             init_specs = spack.cmd.parse_specs(args.constraint)
-            results = env.all_matching_specs(*init_specs)
+            env_specs = env.all_matching_specs(*init_specs)
         else:
-            results = env.all_specs()
+            env_specs = all_env_specs
+
+        spec_hashes = set(x.dag_hash() for x in env_specs)
+        specs_meeting_q_args = set(spack.store.STORE.db.query(hashes=spec_hashes, **q_args))
+
+        results = list()
+        for spec in env_specs:
+            if not spec.installed:
+                concretized_but_not_installed.append(spec)
+            if spec in specs_meeting_q_args:
+                results.append(spec)
     else:
-        q_args = query_arguments(args)
         results = args.specs(**q_args)
 
-    decorator = make_env_decorator(env) if env else lambda s, f: f
-
     # use groups by default except with format.
     if args.groups is None:
         args.groups = not args.format
 
     # Exit early with an error code if no package matches the constraint
-    if not results and args.constraint:
+    if concretized_but_not_installed and args.show_concretized:
+        pass
+    elif results:
+        pass
+    elif args.constraint:
         constraint_str = " ".join(str(s) for s in args.constraint_specs)
         tty.die(f"No package matches the query: {constraint_str}")
```

Comment:
Can you fix this as well then?

```
$ spack find x
==> Error: No package matches the query: x

$ spack -e . find  x
==> Error: No package matches the query: 
```

Reply:
Also better to make it raise than tty.die if you meant to isolate the function for unit tests.